Add SDD for Commodore Compile Pipeline (#181)
projectsyn/documentation: docs/modules/SDDs/pages/0032-compile-pipeline.adoc

= SDD 0031 - Central Component Version tracking

:sdd_author:    Aline Abler

> :sdd_owner:     Project Syn IG

Reviewer: ```suggestion
:sdd_owner:     Project Syn IG
:sdd_reviewers: Simon Gerber
```

It solves the problem of configuration drift, where changes to the tenant repository might not be reflected in every cluster catalog because not all of them have since been compiled, and it lessens the burden on catalog maintainers who otherwise would need to locally compile each cluster individually.

It is already fairly straightforward to manually set up basic auto-compilation for individual tenant repositories without special support from Project Syn itself.
At VSHN, such a solution has been in use for several years.

However, certain features (such as automatic configuration of the compile pipeline) are hard to implement in a standalone fashion.
As such features are now desired, it makes sense to fully integrate the compile pipeline into Project Syn.

By making Project Syn "CI-aware", we can implement more seamless management of compile pipeline configuration on the tenant repositories, including automated setup and automated token rotation.
This will go hand-in-hand with the existing repository management features in Lieutenant.

Currently, we're looking at a solution that is specific to GitLab CI/CD, and to tenant and catalog repositories that are stored on GitLab.

> Other CI solutions and other repository hosts might be supported in the future.

Reviewer: IMO, this paragraph doesn't belong in the motivation. It can go into the "Implementation Details/Notes/Constraints" section.

=== Goals

* Provide pipeline definitions for GitLab CI/CD to automatically compile and push cluster catalogs on a tenant repository.
* Enable Lieutenant to autonomously manage the configuration required to set up the compile pipeline for a cluster catalog.

=== Non-Goals

* Support for CI solutions other than GitLab CI/CD.

== Design Proposal

=== Gitlab CI/CD Pipeline for Commodore

> The pipeline definitions for the Commodore compile pipeline are managed in a dedicated GitLab repository (the pipeline repository).

Reviewer: Afaict, we don't need to manage the included CI config in a GitLab repo (cf. https://docs.gitlab.com/ee/ci/yaml/includes.html#include-a-single-configuration-file), we can import the config from an arbitrary HTTPS URL.

```suggestion
The pipeline definitions for the Commodore compile pipeline are managed in a dedicated Git repository (the pipeline repository).
This repository can be hosted anywhere.
```

> Tenant repositories can then include these pipeline definitions in their `gitlab-ci.yml`.

Reviewer: ```suggestion
Tenant repositories can then include these pipeline definitions in their `gitlab-ci.yml` as a "remote" include.
See the https://docs.gitlab.com/ee/ci/yaml/includes.html#include-a-single-configuration-file[GitLab CI/CD documentation] for details.
```

* `COMMODORE_API_URL`. This contains the URL at which the Lieutenant API can be accessed.
* `COMMODORE_API_TOKEN`. This contains an access token for the Lieutenant API.
* `CLUSTERS`. This contains a space-separated list of cluster IDs which should be compiled and pushed automatically.

=== CRD

We add a new field to the `GitRepoTemplate` (and, by extension, the `GitRepo`) CRD, under the `.spec` key, called `compilePipeline`.

The `compilePipeline` field contains configuration pertaining to the automatic setup of the compile pipeline on the tenant repository.
It is optional.
Absence of the field disables automatic setup and management of the compile pipeline.

The `compilePipeline` field contains a dict with the following fields:

* `gitlabCatalogProjectAccessSecretRef`: (for cluster catalog repos) Reference to a secret containing the GitLab Project Access Token that can be used to access this repository. (Managed by Lieutenant)
* `pipelineFiles`: Dictionary containing file paths as keys, and file contents as values.

>    These files will be committed to the tenant repository when the compile pipeline is configured, if they do not exist already.

Reviewer: Please clarify here that both of these fields are optional and trigger certain actions by the operator when they're present.

> We add a new field to the `Cluster` CRD, under the `.status` key, called `compileMeta`.

Reviewer: Copy-pasta?

```suggestion
```

Author: yes 🙃

[source,yaml]
----
apiVersion: syn.tools/v1alpha1
kind: Cluster
metadata:
  name: my-cluster
spec:
  gitRepoTemplate:
    compilePipeline:
      gitlabCatalogProjectAccessSecretRef: c-my-cluster-project-access
      pipelineFiles:
        .gitlab-ci.yml: |
          include:
            - project: syn/commodore-compile-pipeline
              ref: master

>               file: /.gitlab/commodore-common.yml

Reviewer: I would provide both an example `Tenant` and `Cluster` object here. This example is misleading and doesn't match the explanation of how it's actually supposed to be used in the following section.

----

=== Operator

The Lieutenant Operator will be extended to automatically manage the compile pipeline for repositories where this is enabled (by way of configuring the `compilePipeline` field).

> Since the compile pipeline has to interact with both the tenant repo as well as the cluster catalog repo, it must be enabled on both corresponding `gitRepo` definitions for the configuration to be functioning.

Reviewer: You've used "repo" here, but "repository" elsewhere in the text. I'd normalize to either, but I prefer "repository" for the SDDs.

```suggestion
Since the compile pipeline has to interact with both the tenant repository as well as the cluster catalog repositories, it must be enabled on both corresponding `gitRepo` definitions for the configuration to be functional.
```

When `compilePipeline` is configured on a cluster repository (catalog repository), Lieutenant does the following:

* Create a new `Project Access Token` on the catalog repository via the GitLab API, using the cluster repository's configured apiSecretRef.
* Store the new access token in a new secret.
* Store a reference to the new secret in `catalogProjectAccessSecretRef`.
* Add a new CI/CD variable to the corresponding tenant repository via the GitLab API, using the tenant repository's configured apiSecretRef. The value is named `ACCESS_TOKEN_CLUSTERNAME` (where `CLUSTERNAME` is the cluster ID with `-` replaced by `_`) and contains the GitLab project access token as its value.
* Update the CI/CD variable `CLUSTERS` on the corresponding tenant repository to ensure it includes the cluster's ID.

When `compilePipeline` is configured on a *managed* tenant repository, Lieutenant does the following:

* Render the `pipelineFiles` configured on the tenant repository, and commit them.
* Add a new CI/CD variable `COMMODORE_API_URL` to the tenant repository, containing the API URL for Lieutenant.

> * Add a new CI/CD variable `COMMODORE_API_TOKEN` to the tenant repository, containing the API token for Lieutenant.

Reviewer: ```suggestion
* Add a new CI/CD variable `COMMODORE_API_TOKEN` to the tenant repository, containing the tenant's API token for Lieutenant.
```

When the `compilePipeline` field is removed from a repository, all the corresponding resources should be cleaned up.

Unmanaged repositories cannot have automatically configured compile pipelines.

==== Rotation of Project Access Tokens

GitLab's project access tokens have a limited lifespan.
Lieutenant should automatically rotate them as required.

To this end, a scheduled task should run in the operator that checks the token's expiration date via the GitLab API, and rotates any that are close to expiring.

To rotate a token:

* Use the GitLab API to rotate the existing token.
* Update the corresponding catalog project access secret.
* Update the corresponding CI/CD variable (`ACCESS_TOKEN_CLUSTERNAME`) in the tenant repository.

=== Implementation Details/Notes/Constraints

Existing compile pipeline configuration::
If a setup already includes a bunch of tenant repositories with manually configured CI/CD, some care has to be taken to ensure the new implementation can "adopt" this configuration.
+

> In particular, these repositories would already have a working `.gitlab-ci.yml` that probably can be left as-is.

Reviewer: This may not be the case for our setup with the current proposal, since we currently have the `CLUSTERS` CI variable in the `.gitlab-ci.yml` and not as a CI/CD settings variable. Did you verify which variable value "wins"?

Author: Yes, the CI/CD variables will win. See also https://docs.gitlab.com/ee/ci/variables/#cicd-variable-precedence

At VSHN, such a solution has been in use for several years.

However, certain features (such as automatic configuration of the compile pipeline) are hard to implement in a standalone fashion.
As such features are now desired, it makes sense to fully integrate the compile pipeline into Project Syn.

By making Project Syn "CI-aware", we can implement more seamless management of compile pipeline configuration on the tenant repositories, including automated setup and automated token rotation.
This will go hand-in-hand with the existing repository management features in Lieutenant.

=== Goals

* Provide pipeline definitions for GitLab CI/CD to automatically compile and push cluster catalogs on a tenant repository.
* Enable Lieutenant to autonomously manage the configuration required to set up the compile pipeline for a cluster catalog.

=== Non-Goals

* Support for CI solutions other than GitLab CI/CD.

== Design Proposal

=== Requirements for Pipeline Configuration

Lieutenant imposes certain assumptions on the configuration of the pipeline:
Namely, the pipeline has to be set up on the tenant repository by way of adding (arbitrary) files to the repository,

> and it is configured through setting CI/CD variables on the repository.

Reviewer: ```suggestion
Namely, the pipeline has to be set up on the tenant repository by way of adding (arbitrary) files to the repository, and it is configured through setting CI/CD variables on the repository.
```

In particular, Lieutenant configures the following CI/CD variables:

* `ACCESS_TOKEN_CLUSTERNAME`, where `CLUSTERNAME` is the name of a specific cluster, with `-` replaced by `_`.
  This contains a GitLab Project Access Token, which must have read-write access the corresponding cluster's catalog repository.
* `COMMODORE_API_URL`. This contains the URL at which the Lieutenant API can be accessed.
* `COMMODORE_API_TOKEN`. This contains an access token for the Lieutenant API.
* `CLUSTERS`. This contains a space-separated list of cluster IDs which should be compiled and pushed automatically.

=== GitRepo CRD

We add two new fields to the `GitRepoTemplate` (and, by extension, the `GitRepo`) CRD, under the `.spec` key, called `accessTokenSecretRef` and `ciVariables`.

> The `accessTokenSecretRef` field contains a reference to a secret.

Reviewer: In my mind, `secretRef` has the connotation of referencing a secret that's managed externally. It may be more obvious that this field contains the name of a secret that's generated by the operator if the field is named `accessTokenSecretName`.

> If it is set, the Lieutenant operator will store an access token into this secret, which can be used to access the git repository.

Reviewer: ```suggestion
If it is set, the Lieutenant operator will store an access token into this secret, which can be used to access the Git repository.
```

In the case of GitLab, this would be a Project Access Token with read-write access to the repository.

The `ciVariables` field contains a dictionary describing variable names and corresponding values.

> These variables are added to the git repository as CI/CD variables.

Reviewer: Thinking about this again, it will probably make sense to allow the values to be similar to K8s's `env[*].value` field. That way, we don't need to put secrets verbatim in `spec.ciVariables` but can instead use a K8s secretRef.

kind: GitRepo
metadata:
  name: my-repo
spec:
  accessTokenSecretRef: my-repo-access-token
  ciVariables:
    COMMODORE_API_URL: ...
    COMMODORE_API_TOKEN: ...
----

=== Cluster CRD

We add a new field to the `Cluster` CRD, under the `.spec` key, called `enableCompilePipeline`.

The field contains a boolean flag, which controls whether the compile pipeline should be enabled or disabled for this cluster.

It is optional; not specifying it is equivalent to setting it to `false`.

[source,yaml]
----
apiVersion: syn.tools/v1alpha1
kind: Cluster
metadata:
  name: c-my-cluster
spec:
  enableCompilePipeline: true
----

=== Tenant CRD

We add a new field to the `Tenant` CRD, under the `.spec` key, called `compilePipeline`.

The `compilePipeline` field contains configuration pertaining to the automatic setup of the compile pipeline on the tenant repository.
It is optional.
Absence of the field disables automatic setup and management of the compile pipeline.

The `compilePipeline` field contains a dict with the following fields:

* `clusters`: List of cluster IDs of clusters for which the compile pipeline should be executed.
  This field is managed by the operator.

> * `pipelineFiles`: Dictionary containing file paths as keys, and file contents as values.

Author: I was wondering if it isn't better to just directly add these to `gitRepoTemplate.templateFiles`, I'm not certain what we gain from this extra step. Thoughts?

Reviewer: I was thinking about that as well when writing the initial comment, I guess if we expect the users to provide the complete contents of each pipeline file anyway, then there's no real need for the extra field.

  In the case of GitLab, this is a Project Access Token.
  The operator also runs a scheduled job which refreshes these tokens when they are close to expiring, or when they no longer exist on the repository host.
* The content of `.spec.ciVariables` is written to the repository's configuration on the git host.
  In the case of GitLab, it is written as CI/CD variables.

NOTE: If the GitRepo is of type `unmanaged`, none of these steps will be executed.

The operator will reconcile *Clusters* as follows:

* When `.spec.enableCompilePipeline` is set to `true`, the tenant's `spec.compilePipeline.clusters` is updated to contain the cluster ID.
* Similarly, when the field is set to `false` or missing, the tenant's `spec.compilePipeline.clusters` is updated to not contain the cluster ID.

The operator will reconcile *Tenants* as follows:

* When `.spec.compilePipeline` exists and isn't empty, the following entries are added to the tenant repository GitRepo's `.spec.ciVariables`:
** `COMMODORE_API_URL`, containing the URL at which the Lieutenant API can be accessed.

> ** `COMMODORE_API_TOKEN`, containing the tenant's access token for the Lieutenant API.

Reviewer: ```suggestion
** `COMMODORE_API_TOKEN`, containing a reference to the secret which contains the tenant's access token for the Lieutenant API.
```

** `CLUSTERS`, containing a space-separated list of cluster IDs taken directly from `.spec.compilePipeline.clusters`.
* For each entry in `.spec.compilePipeline.pipelineFiles` whose value isn't `{remove}`, a new corresponding entry is added to the tenant's `.spec.gitRepoTemplate.templateFiles`.

> * For each entry in `.spec.compilePipeline.pipelineFiles` whose value is `{remove}`, the corresponding entry is removed from the tenant's `.spec.gitRepoTemplate.templateFiles`.

Reviewer: This doesn't need to be specified here, since we still need to replicate entries whose value is `{remove}` verbatim to `spec.gitRepoTemplate.templateFiles` in order for the file to actually be deleted in the Git repo.

Author: Oh! I misunderstood how it works! All right then

* For each entry in `.spec.compilePipeline.clusters`, another entry is added to the tenant repository GitRepo's `spec.ciVariabes`.
  The key is `ACCESS_TOKEN_CLUSTERNAME`, where `CLUSTERNAME` is the ID of a specific cluster, with `-` replaced by `_`.

>   The value is the access token to access that cluster's catalog repository, taken from the secret specified in the catalog GitRepo configuration under `.spec.accessTokenSecretRef`.

Reviewer: ```suggestion
  The value is a reference to the secret containing the access token to access that cluster's catalog repository, taken from the secret specified in the catalog GitRepo configuration under `.spec.accessTokenSecretRef`.
```

> = SDD 0031 - Central Component Version tracking

Reviewer 2: ```suggestion
= SDD 0032 - Commodore Compile Pipeline for GitLab
```

:sdd_author:    Aline Abler
:sdd_owner:     Project Syn IG

> :sdd_reviewers: Simon Gerber, Tobias Brunner

Reviewer 2: ```suggestion
:sdd_reviewers: Simon Gerber, Tobias Brunner, Sebastian Widmer
```

:sdd_date:      2024-06-13
:sdd_status:    draft

include::partial$meta-info-table.adoc[]

[NOTE]
.Summary
====

> This describes how we want to implement CI/CD for Project Syn using a Commodore compile pipeline, which enables automatic compilation of cluster catalogs whenever the corresponding tenant repository is modified.

Reviewer 2: ```suggestion
This describes how we want to implement GitLab CI/CD for Project Syn using a Commodore compile pipeline, which enables automatic compilation of cluster catalogs whenever the corresponding tenant repository is modified.
```

Each object in the list has a type that's modeled after a Kubernetes container's `env` field.
In contrast to container environment variables, our variables only support specifying values directly (via field `value`) or by referencing a `Secret` resource (via field `valueFrom.secretKeyRef`).
These variables are added to the Git repository as CI/CD variables.

[source,yaml]
----
apiVersion: syn.tools/v1alpha1
kind: GitRepo
metadata:
  name: my-repo
spec:
  accessTokenSecretName: my-repo-access-token
  ciVariables:
    - name: COMMODORE_API_URL
      value: ...
    - name: COMMODORE_API_TOKEN
      valueFrom:
        secretKeyRef:
          name: api-token-secret
          key: token
----

=== Cluster CRD

We add a new field to the `Cluster` CRD, under the `.spec` key, called `enableCompilePipeline`.

The field contains a boolean flag, which controls whether the compile pipeline should be enabled or disabled for this cluster.

It is optional; not specifying it is equivalent to setting it to `false`.

[source,yaml]
----
apiVersion: syn.tools/v1alpha1
kind: Cluster
metadata:
  name: c-my-cluster
spec:

>   enableCompilePipeline: true

Reviewer 2: Do we want to make it more explicit as in `enableGitLabCompilePipeline` or are we assuming future implementations for GitHub or others also use this flag?

Author: From how it's currently designed, I can easily envision future versions using the same flag. We already have a way to distinguish between different git hosts in the gitrepo spec. The rest of the config should be transferable.

** `COMMODORE_API_URL`, containing the URL at which the Lieutenant API can be accessed.
** `COMMODORE_API_TOKEN`, containing a reference to the secret which contains the tenant's access token for the Lieutenant API.
** `CLUSTERS`, containing a space-separated list of cluster IDs taken directly from `.spec.compilePipeline.clusters`.
* For each entry in `.spec.compilePipeline.clusters`, another entry is added to the tenant repository GitRepo's `spec.ciVariabes`.
  The key is `ACCESS_TOKEN_CLUSTERNAME`, where `CLUSTERNAME` is the ID of a specific cluster, with `-` replaced by `_`.
  The value is a reference to the secret containing the access token to access that cluster's catalog repository, taken from the secret specified in the catalog GitRepo configuration under `.spec.accessTokenSecretName`.
* For each entry in `.spec.compilePipeline.pipelineFiles`, a new corresponding entry is added to the tenant's `.spec.gitRepoTemplate.templateFiles`.

=== Implementation Details/Notes/Constraints

Currently, we're looking at a solution that is specific to GitLab CI/CD, and to tenant and catalog repositories that are stored on GitLab.

> Other CI solutions and other repository hosts might be supported in the future.

Reviewer 2: ```suggestion
Other CI solutions and other repository hosts might be supported in the future.
The implementation takes care to gracefully skip the CI steps if not on GitLab.
We leave the implementation extensible enough to add further Git management tools and CI systems without needing breaking changes.
```

Existing compile pipeline configuration::

> If a setup already includes a bunch of tenant repositories with manually configured CI/CD, some care has to be taken to ensure the new implementation can "adopt" this configuration.

Reviewer 2: I don't necessarily agree that we want/need to adopt anything. I'd rather have consistency and i don't see how having a clean cut would break something.

Author: Fair enough. At the moment, it's written in a way where adoption will "just work" except perhaps in some exceptional cases where the in-repo config has been manually verbaschtled.
I agree it's not useful to try and account for those latter cases.

Author: I've rephrased the section a bit to not make it seem like we want to implement extrawurst support for all sorts of existing manual config.

= SDD 0031 - Central Component Version tracking

:sdd_author:    Aline Abler
:sdd_owner:     Project Syn IG

> :sdd_reviewers: Simon Gerber

Reviewer: ```suggestion
:sdd_reviewers: Simon Gerber, Tobias Brunner
```

In particular, Lieutenant configures the following CI/CD variables:

* `ACCESS_TOKEN_CLUSTERNAME`, where `CLUSTERNAME` is the name of a specific cluster, with `-` replaced by `_`.
  This contains a GitLab Project Access Token, which must have read-write access the corresponding cluster's catalog repository.
* `COMMODORE_API_URL`. This contains the URL at which the Lieutenant API can be accessed.
* `COMMODORE_API_TOKEN`. This contains an access token for the Lieutenant API.
* `CLUSTERS`. This contains a space-separated list of cluster IDs which should be compiled and pushed automatically.

=== GitRepo CRD

We add two new fields to the `GitRepoTemplate` (and, by extension, the `GitRepo`) CRD, under the `.spec` key, called `accessTokenSecretName` and `ciVariables`.

The `accessTokenSecretName` field contains a reference to a secret.
If it is set, the Lieutenant operator will store an access token into this secret, which can be used to access the Git repository.
In the case of GitLab, this would be a Project Access Token with read-write access to the repository.

> The `ciVariables` field contains a dictionary describing variable names and corresponding values.

Reviewer: With the latest changes, this sentence doesn't match the example anymore.

```suggestion
The `ciVariables` field contains a list of objects describing variable names and corresponding values.
Each object in the list has a type that's modeled after a Kubernetes container's `env` field.
In contrast to container environment variables, our variables only support specifying values directly (via field `value`) or by referencing a `Secret` resource (via field `valueFrom.secretKeyRef`).
```

Not sure if we should explicitly allow configmaps as source in addition to secrets

> These variables are added to the git repository as CI/CD variables.

Reviewer: Afaik, the Git project spells itself with a capital G.

```suggestion
These variables are added to the Git repository as CI/CD variables.
```

  ciVariables:
    - name: COMMODORE_API_URL
      value: ...
    - name: COMMODORE_API_TOKEN
      valueFrom:
        secretKeyRef:
          name: api-token-secret
          key: token
----

=== Cluster CRD

We add a new field to the `Cluster` CRD, under the `.spec` key, called `enableCompilePipeline`.

The field contains a boolean flag, which controls whether the compile pipeline should be enabled or disabled for this cluster.

It is optional; not specifying it is equivalent to setting it to `false`.

[source,yaml]
----
apiVersion: syn.tools/v1alpha1
kind: Cluster
metadata:
  name: c-my-cluster
spec:
  enableCompilePipeline: true
----

=== Tenant CRD

We add a new field to the `Tenant` CRD, under the `.spec` key, called `compilePipeline`.

The `compilePipeline` field contains configuration pertaining to the automatic setup of the compile pipeline on the tenant repository.
It is optional.
Absence of the field disables automatic setup and management of the compile pipeline.

The `compilePipeline` field contains a dict with the following fields:

* `clusters`: List of cluster IDs of clusters for which the compile pipeline should be executed.

>   This field is managed by the operator.

Reviewer: Would it make sense to move this to `status.compilePipeline` now that it's intended to be fully operator-managed?

Author: Since we're going back to having the `pipelineFiles`, it doesn't make sense to fully move this into status anymore. But we might want to split it into configurable vs. operator managed fields and only put the latter in status. I'm not sure what's better practice - keep them together, or have a strict split between managed and user-defined?

Reviewer: I'm not sure what the best practice is here, maybe @bastjan can chime in?

Reviewer 2: I prefer fully operator managed fields in `.status` it removes a source of misunderstandings. Not a super firm position since Kubernetes also fully manages a lot of fields in `.spec` in the core manifests.

Author: I've split it up now. This prompted me to also add an `enabled` field in the spec.

[source,yaml]
----
apiVersion: syn.tools/v1alpha1
kind: Tenant
metadata:
  name: t-my-tenant
spec:
  compilePipeline:
    clusters:
      - c-my-cluster
----

=== In-Repo CI/CD pipeline configuration

Configuring the CI pipeline usually happens through files committed to the corresponding repository.
For a Lieutenant-managed pipeline configuration, these files should be managed by Lieutenant.

> To achieve this, we can leverage the existing mechanism to commit template files to git repositories:

Reviewer: ```suggestion
To achieve this, we can leverage the existing mechanism to commit template files to Git repositories:
```

I've looked at that again, and have discovered that Lieutenant will unconditionally set any entries in `templateFiles` to `{remove}` unless the file name matches a currently existing cluster (`c-existing-cluster-id.yml`) or `common.yml`. Therefore, to make this work we'll need to either adjust Lieutenant to allow "unmanaged" entries in `templateFiles` or we'll need to go back to the initial approach with an additional new field in `spec.compilePipeline` where we can define CI pipeline config files.

See https://github.com/projectsyn/lieutenant-operator/blob/831dba95bb3aa4e6dc83fea70e6ecd0f57fa1bd8/controllers/tenant/git.go#L14-L60 for the current implementation.

Author: I decided to go back to the old way of having `compilePipeline.pipelineFiles` in the tenant spec.

Small bonus: since the operator will autoremove files that aren't either in the pipeline or have the cluster ID as a name (or are `common.yml`), there will no longer be a need for the user to explicitly set `{remove}`.

[source,yaml]
----
apiVersion: syn.tools/v1alpha1
kind: Tenant
metadata:
  name: t-my-tenant
spec:
  gitRepoTemplate:
    templateFiles:
      .gitlab-ci.yml: |
        include:
          - project: syn/commodore-compile-pipeline
            ref: master
            file: /.gitlab/commodore-common.yml

----

=== Operator

> The Lieutenant Operator will be extended to automatically manage the compile pipeline for repositories where this is enabled (by way of configuring the `compilePipeline` field on the tenant and the `enableCompilePipeline` field on the cluster).

Reviewer: ```suggestion
The Lieutenant Operator will be extended to automatically manage the compile pipeline for repositories where this is enabled (by way of deploying the CI config file in the tenant and the `enableCompilePipeline` field on the cluster).
```

Since the compile pipeline has to interact with both the tenant repository as well as the cluster catalog repositories, it must be enabled on both corresponding resources for the configuration to be functional.
This way, it is possible to enable auto-compilation for some, but not all clusters on a tenant.

The operator will reconcile *GitRepos* as follows:

* When `spec.accessTokenSecretName` is set, the operator generates an access token for the corresponding repository (via the repository host's API, using the API secret in `.spec.apiSecretRef`), and writes this token into a secret with the given name.
  In the case of GitLab, this is a Project Access Token.
  The operator also runs a scheduled job which refreshes these tokens when they are close to expiring, or when they no longer exist on the repository host.

> * The content of `.spec.ciVariables` is written to the repository's configuration on the git host.

Reviewer: ```suggestion
* The content of `.spec.ciVariables` is written to the repository's configuration on the Git host.
```

= SDD 0032 - Commodore Compile Pipeline

:sdd_author:    Aline Abler
:sdd_owner:     Project Syn IG
:sdd_reviewers: Simon Gerber, Tobias Brunner, Sebastian Widmer
:sdd_date:      2024-06-13

> :sdd_status:    draft

Reviewer: ```suggestion
:sdd_status:    accepted
```

====

== Motivation

Having a continuous integration solution unlocks a number of benefits:
It solves the problem of configuration drift, where changes to the tenant repository might not be reflected in every cluster catalog because not all of them have since been compiled, and it lessens the burden on catalog maintainers who otherwise would need to locally compile each cluster individually.

It is already fairly straightforward to manually set up basic auto-compilation for individual tenant repositories without special support from Project Syn itself.
At VSHN, such a solution has been in use for several years.

However, certain features (such as automatic configuration of the compile pipeline) are hard to implement in a standalone fashion.
As such features are now desired, it makes sense to fully integrate the compile pipeline into Project Syn.

By making Project Syn "CI-aware", we can implement more seamless management of compile pipeline configuration on the tenant repositories, including automated setup and automated token rotation.
This will go hand-in-hand with the existing repository management features in Lieutenant.

=== Goals

> * Provide pipeline definitions for GitLab CI/CD to automatically compile and push cluster catalogs on a tenant repository.

Reviewer: This is actually not really achieved anymore with the latest iteration. I'd move this to non-goals for this SDD since we now exclude pipeline definitions completely, and replace it with the following

```suggestion
* Specify the interface that Commodore CI pipeline definitions need to implement in order to work with the managed CI/CD configuration
```

Author: Well spotted 👍🏻

* Enable Lieutenant to autonomously manage the configuration required to set up the compile pipeline for a cluster catalog.

=== Non-Goals

> * Support for CI solutions other than GitLab CI/CD.

Reviewer: ```suggestion
* Support for CI solutions other than GitLab CI/CD.
* Provide pipeline definitions to automatically compile and push cluster catalogs on a tenant repository.
```
